New rule: Text nodes have minimal contrast (#833)
act-rules/act-rules.github.io: _rules/text-node-contrast-afw4f7.md

---
id: afw4f7
name: Text node has minimal contrast
rule_type: atomic

description: |
  This rule checks that every text node has minimal color contrast with its background

accessibility_requirements:

>   wcag20:1.4.3: # Contrast (Minimum)

Reviewer: may be it is representation in github, but these look like `tabs` over `spaces`

Author: Not seeing it. Resolve

    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed

input_aspects:
  - DOM Tree
  - CSS Styling
  - Language

authors:
  - Brian Bors
  - Kasper Isager
  - Wilco Fiers
---

## Applicability

> Any [text node](https://dom.spec.whatwg.org/#text) that is [visible](#visible) and a [child](https://dom.spec.whatwg.org/#concept-tree-child) (in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree)) of an HTML element, except if the [text node](https://dom.spec.whatwg.org/#text) is a [descendent](https://dom.spec.whatwg.org/#concept-shadow-including-descendant) of an element that:

Reviewer 2: drafts.csswg.org is WiP and should be cited as such.

- Has a [semantic role](#semantic-role) that inherits from [widget](https://www.w3.org/TR/wai-aria-1.1/#widget)
- Is used in the [accessible name](#accessible-name) of a [widget](https://www.w3.org/TR/wai-aria-1.1/#widget) that is [disabled](#disabled) (i.e. It is part of the label)
- Has a [semantic role](#semantic-role) of [group](https://www.w3.org/TR/wai-aria-1.1/#group) that is [disabled](#disabled)

> **Note**: When the text color is the same as the background color, the element is not visible, and so is not applicable.

Reviewer 2: I think this note should be moved to the definition of `visible`.

Also, most of our "invisible" examples (especially when it need to be in the accessibility tree) in rules (all rules) tend to be with text out of screen (`style="position: absolute; top: -9999px; left: -9999px;"`). We should probably add some with text color being same as background color.

Reviewer 3: The definition of visible usually has nothing to with differences in colour so it should not be mentioned in that definition.

Reviewer 2: @ShadowBB I do not fully understand. Do you mean that
```html
<p style="color: #000; background: #000;">Some text</p>
```
This is visible and thus should be checked by all rules that have "visible" in the applicability?

## Expectation

For each test target, the [highest possible contrast](#highest-possible-contrast) between the [foreground colors](#foreground-colors-of-text) and [background colors](#background-colors-of-text) is at least 4.5:1 or 3.0:1 for [larger scale text](#large-scale-text), except if the [text node](https://dom.spec.whatwg.org/#text) is [decorative](#decorative), or doesn't express anything in [human language](https://www.w3.org/TR/WCAG21/#dfn-human-language-s).

> **Note**: Passing this rule does not mean that the [text node](https://dom.spec.whatwg.org/#text) has sufficient color contrast. White text on a black and white image passes this rule, whether or not that text is legible depends on the how much of the white words are positioned on the black parts of the image. This requires further testing.

Reviewer: I think we are trying to say we do not consider other composited elements/ nodes that may directly lie under a given text node. Just saying a white text over a black or white image is not considered, tells me if I place a pseudo style or an absolutely positioned black box under a black text node, then this rule will pass. This note should be written in a more generic fashion perhaps.

## Assumptions

> Success criterion 1.4.3 has exceptions for "incidental" text, which includes inactive user interface components and decorative texts. The rule assumes that [text nodes](https://dom.spec.whatwg.org/#text) should be ignored are hidden from assistive technologies. If this isn't the case, the rule may produce incorrect results.

Reviewer 2: "The rule assumes that text nodes should…" => "that text nodes **that** should…"

> Success criterion 1.4.3 also has an exception for logos and brand names. Since logos and brand names are usually displayed through images to ensure correct rendering, this rule does not take logos or brand names into consideration. If a logo or brand name is included using [text nodes](https://dom.spec.whatwg.org/#text), this rule may not produce the rule may produce incorrect results.

Reviewer 2: "this rule may not produce the rule may produce " => choose one.

## Accessibility Support

> Different browsers have different levels of support for CSS. This can cause contrast issues in one browser that do not appear in another. Because of that, this rule can produce different results depending on the browser that is used.

Reviewer: Different levels of CSS support based on browser? Perhaps narrow it with an example of any color related property that are inconsistent across browsers?

## Background

- [Understanding Success Criterion 1.4.3: Contrast (Minimum)](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html)
- [G18: Ensuring that a contrast ratio of at least 4.5:1 exists between text (and images of text) and background behind the text](https://www.w3.org/WAI/WCAG21/Techniques/general/G18)
- [G145: Ensuring that a contrast ratio of at least 3:1 exists between text (and images of text) and background behind the text](https://www.w3.org/WAI/WCAG21/Techniques/general/G145)
- [F83: Failure of Success Criterion 1.4.3 and 1.4.6 due to using background images that do not provide sufficient contrast with foreground text (or images of text)](https://www.w3.org/WAI/WCAG21/Techniques/failures/F83)

> ## Test Cases

Reviewer 2: Can you turn examples description into actual sentences (_i.e._ with a verb)?

Reviewer 3: Done thanks!

### Passed

#### Passed Example 1

Black [text node](https://dom.spec.whatwg.org/#text) on a white background.

```html

> <p style="color: #333; background: #FFF;">

Reviewer: perhaps an example with a play in CSS specificity?

	Some text in English
</p>
```

#### Passed Example 2

Black [text node](https://dom.spec.whatwg.org/#text) on a partially white gradient background.

```html
<p style="color: #333; background: linear-gradient(to right, #FFF, #00F);">
	Some text in English
</p>
```

#### Passed Example 3

Light [text node](https://dom.spec.whatwg.org/#text) on a partially black background image.

```html
<p

> 	style="color: #CCC; height:50px; padding-top:25px; background: #000 no-repeat -20px -20px url('../test-assets/black-hole.jpeg');"

Reviewer: can we add the test-assets under a specific directory pertaining to the ruleId?

eg: `test-assets/XXXXXX/something.xyz`

>
	Black hole sun
</p>
```

#### Passed Example 4

[text node](https://dom.spec.whatwg.org/#text) that passes because of a text shadow.

```html
<p style="color: #000; background: #737373; text-shadow: white 0 0 3px">
	Some text in English
</p>
```

#### Passed Example 5

18 point text with a 3:1 contrast.

```html
<p style="color: #000; font-size:18pt; background: #666;">
	Some text in English
</p>
```

#### Passed Example 6

14 point bold text with a 3:1 contrast.

```html
<p style="color: #000; font-size:14pt; font-weight:700; background: #666;">
	Some text in English
</p>
```

#### Passed Example 7

Decorative text with insufficient contrast.

```html
<p>Helvetica is a widely used sans-serif typeface developed in 1957 by Max Miedinger and Eduard Hoffmann.</p>
<p style="font-family: helvetica; background: #EEE; color: #777;" aria-hidden="true">
	The quick brown fox jumps over the lazy dog.
</p>
```

#### Passed Example 8

Text that does not express anything in human language.

```html
<p style="color: #000; background: #666;">

> 	----=====++++++++___________***********%%%%%%%%%%%±±±±@@@@@@@@

Reviewer 2: I kinda guess that the language is an input aspect because "human language" mean "of the page" rather than "any human language that exist(ed) on Earth".
In that case, I think that there should be an example with, say, a page in English and some poorly contrasted text in another language.

Also beware of "reversed spellings of ancient Chinese emperors" 🤣 
http://www.galactanet.com/comic/view.php?strip=284

</p>
```

#### Passed Example 9

Text that is in a shadow DOM tree with sufficient contrast.

```html
<p style="color: #333; background: #fff;" id="p"></p>
<script>
  const shadowRoot = document.getElementById('p').attachShadow({ mode: 'open' })
  shadowRoot.textContent = 'Some text in English'
</script>
```

### Failed

#### Failed Example 1

> Text with insufficient contrast on a plain background.

Reviewer 2: Wondering if there should be a large text failed example also.
And maybe a 4:1 (or something) ratio on small text failed example.

```html
<p style="color: #AAA; background: white;">
	Some text in English
</p>
```

#### Failed Example 2

Text with insufficient contrast on a gradient background.

```html
<p style="color: #AAA; background: linear-gradient(to right, #FFF, #00F);">
	Some text in English
</p>
```

#### Failed Example 3

Text with insufficient contrast on a background image.

```html
<p

> 	style="color: #555; height:50px; padding-top:25px; background: black no-repeat -20px -20px url('../test-assets/black-hole.jpeg');"

Reviewer: can we add the test-assets under a specific directory pertaining to the ruleId?

eg: `test-assets/XXXXXX/something.xyz`

>
	Black hole sun
</p>
```

#### Failed Example 4

Text that fails because of alpha transparancy.

```html
<p style="color: rgba(0,0,0,.3); background: #FFF">
	Some text in English
</p>
```

#### Failed Example 5

Text that fails because of CSS transparent.

```html
<div style="background: #FFF">
	<p style="color: #000; opacity: .3">
		Some text in English
	</p>
</div>
```

#### Failed Example 6

Text that is in a shadow DOM tree with insufficient contrast.

```html
<p style="color: #aaa; background: #fff;" id="p"></p>
<script>
  const shadowRoot = document.getElementById('p').attachShadow({ mode: 'open' })
  shadowRoot.textContent = 'Some text in English'
</script>
```

### Inapplicable

#### Inapplicable Example 1

Invisible text.

```html
<p style="display: none">Some invisible text in English</p>
```

#### Inapplicable Example 2

Text with the same foreground and background colors.

```html
<p style="color: white; background: white;">Some white on white text in English</p>
```

> 

Reviewer 2: Should we add an example where text is invisible because of positioning (the usual "invisible but in accessibility tree" example)?

#### Inapplicable Example 3

Text not a child of an HTML element.

```html
<svg>
	<text x="0" y="15">I love SVG!</text>
</svg>
```

#### Inapplicable Example 4

Not a [text node](https://dom.spec.whatwg.org/#text) but an image.

```html
<p>
	<img

> 		scr="../test-assets/example.png"

Reviewer: can we add the test-assets under a specific directory pertaining to the ruleId?

eg: `test-assets/XXXXXX/something.xyz`

---
id: afw4f7
name: Text node has minimal contrast
rule_type: atomic

description: |
  This rule checks that the highest possible contrast of every text node with its background meets the minimal contrast requirement

accessibility_requirements:
  wcag20:1.4.3: # Contrast (Minimum)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed

input_aspects:
  - DOM Tree
  - CSS Styling
  - Language

authors:
  - Brian Bors
  - Kasper Isager
  - Wilco Fiers
---

## Applicability

Any [text node](https://dom.spec.whatwg.org/#text) that is [visible](#visible) and a [child](https://dom.spec.whatwg.org/#concept-tree-child) (in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree)) of an HTML element, except if the [text node](https://dom.spec.whatwg.org/#text) is a [descendant](https://dom.spec.whatwg.org/#concept-shadow-including-descendant) of an element that:

- Has a [semantic role](#semantic-role) that inherits from [widget](https://www.w3.org/TR/wai-aria-1.1/#widget)

> - Is used in the [accessible name](#accessible-name) of a [widget](https://www.w3.org/TR/wai-aria-1.1/#widget) that is [disabled](#disabled) (i.e. It is part of the label)

Reviewer 2: Then the use of the word "label" here is pretty confusing. It's not a `label` element, it's not what WCAG calls "label", it's yet some third thing that is called "label" 😕 
The fact that it's used without any definition or anything doesn't help. It's supposed to be part of an explanation, but it just doesn't explain anything. I suggest ditching out the full parenthesis here.

> - Is used in the [accessible name](#accessible-name) of a [widget](https://www.w3.org/TR/wai-aria-1.1/#widget) that is [disabled](#disabled) (i.e. It is part of the label)

Reviewer 2: Digging into it it seems that "i.e." and "e.g." should be italicised in French but not in English…
My bad. Sorry for the confusion.

- Has a [semantic role](#semantic-role) of [group](https://www.w3.org/TR/wai-aria-1.1/#group) that is [disabled](#disabled)

**Note**: When the text color is the same as the background color, the element is not visible, and so is not applicable.

## Expectation

For each test target, the [highest possible contrast](#highest-possible-contrast) between the [foreground colors](#foreground-colors-of-text) and [background colors](#background-colors-of-text) is at least 4.5:1 or 3.0:1 for [larger scale text](#large-scale-text), except if the [text node](https://dom.spec.whatwg.org/#text) is [decorative](#decorative), or doesn't express anything in [human language](https://www.w3.org/TR/WCAG21/#dfn-human-language-s).

> **Note**: Passing this rule does not mean that the [text node](https://dom.spec.whatwg.org/#text) has sufficient color contrast. For example: White text on a black and white image passes this rule as long as it touches at least 1 black pixel. Whether or not that text is legible depends on the how much of the white words are positioned on the black parts of the image. This requires further testing.

Reviewer 2: Given how much misunderstanding happened around this rule, even within our small community, I'm in favour of expanding this note to explain the rational behind the choice of the rule.
Namely, that if the highest possible contrast is too small, then it is guaranteed that no contrast is good and the fail-on-fail relationship makes it hard to test further.

Author: I've updated the note, please have another read.

## Assumptions

> Success criterion 1.4.3 has exceptions for "incidental" text, which includes inactive user interface components and decorative texts. The rule assumes that [text nodes](https://dom.spec.whatwg.org/#text) that should be ignored are hidden from assistive technologies. If this isn't the case, the rule may produce incorrect results.

Reviewer 2: ```suggestion
- [Success criterion 1.4.3: Contrast (Minimum)](https://www.w3.org/TR/WCAG21/#contrast-minimum) has exceptions for "incidental" text, which includes inactive user interface components and decorative texts. The rule assumes that [text nodes](https://dom.spec.whatwg.org/#text) that should be ignored are [disabled](#disabled) or hidden from assistive technologies. If this isn't the case, the rule may produce incorrect results.
```
Not entirely sure about the "disabled" part, but that seems to fit both the applicability and the "inactive" part of the exception.

> Success criterion 1.4.3 also has an exception for logos and brand names. Since logos and brand names are usually displayed through images to ensure correct rendering, this rule does not take logos or brand names into consideration. If a logo or brand name is included using [text nodes](https://dom.spec.whatwg.org/#text), this rule may produce incorrect results.

Reviewer 2: ```suggestion
- [Success criterion 1.4.3: Contrast (Minimum)](https://www.w3.org/TR/WCAG21/#contrast-minimum) also has an exception for logos and brand names. Since logos and brand names are usually displayed through images to ensure correct rendering, this rule does not take logos or brand names into consideration. If a logo or brand name is included using [text nodes](https://dom.spec.whatwg.org/#text), this rule may produce incorrect results.
```

## Background

- [Understanding Success Criterion 1.4.3: Contrast (Minimum)](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html)
- [G18: Ensuring that a contrast ratio of at least 4.5:1 exists between text (and images of text) and background behind the text](https://www.w3.org/WAI/WCAG21/Techniques/general/G18)
- [G145: Ensuring that a contrast ratio of at least 3:1 exists between text (and images of text) and background behind the text](https://www.w3.org/WAI/WCAG21/Techniques/general/G145)
- [F83: Failure of Success Criterion 1.4.3 and 1.4.6 due to using background images that do not provide sufficient contrast with foreground text (or images of text)](https://www.w3.org/WAI/WCAG21/Techniques/failures/F83)
- [CSS Scoping Module Level 1 (Editor's Draft)](https://drafts.csswg.org/css-scoping/)

## Test Cases

### Passed

#### Passed Example 1

> This black [text node](https://dom.spec.whatwg.org/#text) is on a white background.

Reviewer 2: ```suggestion
This dark grey [text node](https://dom.spec.whatwg.org/#text) is on a white background.
```

```html
<p style="color: #333; background: #FFF;">
	Some text in a human language
</p>
```

#### Passed Example 2

> This black [text node](https://dom.spec.whatwg.org/#text) is on a partially white gradient background.

Reviewer 2: Maybe add here that the contrast varies from `#00F/#333` to `#333/#FFF` hence the highest possible contrast is `#333/#FFF` which is good. This will illustrate the "highest possible contrast" thingie and hopefully make things more understandable.

> This black [text node](https://dom.spec.whatwg.org/#text) is on a partially white gradient background.

Reviewer 2: ```suggestion
This dark grey [text node](https://dom.spec.whatwg.org/#text) is on a partially white gradient background.
```

```html
<p style="color: #333; background: linear-gradient(to right, #FFF, #00F); width: 300px;">
	Some text in a human language
</p>
```

#### Passed Example 3

> This light [text node](https://dom.spec.whatwg.org/#text) is on a partially black background image.

Reviewer 2: ```suggestion
This light grey [text node](https://dom.spec.whatwg.org/#text) is on a partially black background image.
```

>
	Black hole sun
</p>
```

#### Failed Example 4

This text fails because the alpha transparency significantly lowers the contrast of the otherwise black text.

```html
<p style="color: rgba(0,0,0,.3); background: #FFF">
	Some text in English
</p>
```

#### Failed Example 5

This text fails because the CSS opacity property significantly lowers the contrast of the otherwise black text.

```html
<div style="background: #FFF">
	<p style="color: #000; opacity: .3">
		Some text in English
	</p>
</div>
```

#### Failed Example 6

Even though the text is in a shadow DOM tree, the [text node](https://dom.spec.whatwg.org/#text) is a child of an HTML in the flat tree and has insufficient contrast

```html
<p style="color: #aaa; background: #fff;" id="p"></p>
<script>
	const shadowRoot = document.getElementById('p').attachShadow({ mode: 'open' })
	shadowRoot.textContent = 'Some text in English'
</script>
```

### Inapplicable

#### Inapplicable Example 1

This is invisible text.

```html
<p style="display: none">Some invisible text in English</p>
```

#### Inapplicable Example 2

> The text is inapplicable because it is positioned off screen.

Reviewer 2: ```suggestion
The text is inapplicable because it is positioned off screen, hence not [visible](#visible).
```

```html
<p style="position:absolute; top: -999em">Some invisible text in English</p>
```

#### Inapplicable Example 3

> This text has the same foreground and background colors.

Reviewer 2: ```suggestion
This text has the same foreground and background colors, hence it is not [visible](#visible).
```

```html
<p style="color: white; background: white;">Some white on white text in English</p>
```

#### Inapplicable Example 4

This text is not the child of an HTML element.

```html
<svg>
	<text x="0" y="15">I love SVG!</text>
</svg>
```

#### Inapplicable Example 5

This is not a [text node](https://dom.spec.whatwg.org/#text) but an image.

```html
<p>
	<img scr="../test-assets/contrast/example.png" alt="example" />
</p>
```

#### Inapplicable Example 6

This text is contained in a native button.

```html
<button>My button!</button>
```

#### Inapplicable Example 7

This text is contained in an ARIA button.

```html
<div role="button">My button!</div>
```

#### Inapplicable Example 8

> This text is in a label of a disabled native widget.

Reviewer 2: ```suggestion
This text is in a label of a [disabled](#disabled) native widget.
```

```html
<label style="color:#888; background: white;">
	My name
	<input disabled />
</label>
```

#### Inapplicable Example 9

> This text is in a label of a disabled ARIA widget.

Reviewer 2: ```suggestion
This text is in a label of a [disabled](#disabled) ARIA widget.
```

```html
<label id="my_pets_name" style="color:#888; background: white;">
	My pet's name
</label>
<div
	role="textbox"
	aria-labelledby="my_pets_name"
	aria-disabled="true"
	style="height:20px; width:100px; border:1px solid black;"
>
	test
</div>
```

#### Inapplicable Example 10

> This text is in a disabled fieldset.

Reviewer 2: ```suggestion
This text is in a [disabled](#disabled) fieldset.
```

```html
<fieldset disabled style="color:#888; background: white;">
	<label>
		My name
		<input />
	</label>
</fieldset>
```

#### Inapplicable Example 11

> This text is in a disabled ARIA group.

Reviewer 2: ```suggestion
This text is in a [disabled](#disabled) ARIA group.
```

---
id: afw4f7
name: Text node has minimal contrast
rule_type: atomic

description: |
  This rule checks that every text node has minimal color contrast with its background

accessibility_requirements:
  wcag20:1.4.3: # Contrast (Minimum)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed

input_aspects:
  - DOM Tree
  - CSS Styling
  - Language

authors:
  - Brian Bors
  - Kasper Isager
  - Wilco Fiers
---

## Applicability

> Any [text node](https://dom.spec.whatwg.org/#text) that is [visible](#visible) and a [child](https://dom.spec.whatwg.org/#concept-tree-child) (in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree)) of an HTML element, except if the [text node](https://dom.spec.whatwg.org/#text) is a [descendant](https://dom.spec.whatwg.org/#concept-shadow-including-descendant) of an element that:

Reviewer 4: https://drafts.csswg.org/css-scoping/#flat-tree is from an Editors draft but is my understanding that there is no other definition available.

Reviewer 3: This is correct.

**Note**: Passing this rule does not mean that the [text node](https://dom.spec.whatwg.org/#text) has sufficient color contrast. For example: White text on a black and white image passes this rule as long as it touches at least 1 black pixel. Whether or not that text is legible depends on the how much of the white words are positioned on the black parts of the image. This requires further testing.

## Assumptions

Success criterion 1.4.3 has exceptions for "incidental" text, which includes inactive user interface components and decorative texts. The rule assumes that [text nodes](https://dom.spec.whatwg.org/#text) that should be ignored are hidden from assistive technologies. If this isn't the case, the rule may produce incorrect results.

> Success criterion 1.4.3 also has an exception for logos and brand names. Since logos and brand names are usually displayed through images to ensure correct rendering, this rule does not take logos or brand names into consideration. If a logo or brand name is included using [text nodes](https://dom.spec.whatwg.org/#text), this rule may produce incorrect results.

Reviewer 3: The note is not about logos and the note says that sometimes this rule passes while the text contrast is still to low. While this assumption tells us of a scenario where this rule would judge the contrast too low while it should actually pass. So we need both.

## Background

- [Understanding Success Criterion 1.4.3: Contrast (Minimum)](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html)
- [G18: Ensuring that a contrast ratio of at least 4.5:1 exists between text (and images of text) and background behind the text](https://www.w3.org/WAI/WCAG21/Techniques/general/G18)
- [G145: Ensuring that a contrast ratio of at least 3:1 exists between text (and images of text) and background behind the text](https://www.w3.org/WAI/WCAG21/Techniques/general/G145)
- [F83: Failure of Success Criterion 1.4.3 and 1.4.6 due to using background images that do not provide sufficient contrast with foreground text (or images of text)](https://www.w3.org/WAI/WCAG21/Techniques/failures/F83)
- [CSS Scoping Module Level 1 (Editor's Draft)](https://drafts.csswg.org/css-scoping/)

## Test Cases

### Passed

#### Passed Example 1

This black [text node](https://dom.spec.whatwg.org/#text) is on a white background.

```html
<p style="color: #333; background: #FFF;">
	Some text in a human language
</p>
```

#### Passed Example 2

> This black [text node](https://dom.spec.whatwg.org/#text) is on a partially white gradient background.

Reviewer 3: No. This is recently changed. As per proposal form @Jym77 .

```html

> <p style="color: #333; background: linear-gradient(to right, #FFF, #00F);">

Author: This needs to have a width on it to ensure the text never gets behind the darker blue.

	Some text in a human language
</p>
```

#### Passed Example 3

This light [text node](https://dom.spec.whatwg.org/#text) is on a partially black background image.

```html
<p
	style="color: #CCC; height:50px; padding-top:25px; background: #000 no-repeat -20px -20px url('../test-assets/contrast/black-hole.jpeg');"
>
	Black hole sun
</p>
```

#### Passed Example 4

This [text node](https://dom.spec.whatwg.org/#text) passes because of a text shadow.

```html
<p style="color: #000; background: #737373; text-shadow: white 0 0 3px">
	Some text in a human language
</p>
```

#### Passed Example 5

This text only has a 3:1 contrast but it is a large (18 point) text so it still passes.

```html
<p style="color: #000; font-size:18pt; background: #666;">
	Some text in a human language
</p>
```

#### Passed Example 6

This text only has a 3:1 contrast but it is a large (14 points and bold) text so it still passes.

```html
<p style="color: #000; font-size:14pt; font-weight:700; background: #666;">
	Some text in English
</p>
```

#### Passed Example 7

> This text has an insufficient contrast but it is decorative so it still passes.

Reviewer 5: But if you look at ACT-R glossary entry for decorative it reads "Serving only an aesthetic purpose, providing no information, and having no functionality."
I would argue that the text in this example meets this definition. But it would be nice to have decorative linked to its definition.

> This text has an insufficient contrast but it is decorative so it still passes.

Author: @Brynanders The reason this text is decorative is because if this text was completely replaced with another sentence, nothing is lost. The purpose of this text is to demonstrate the font, any other sentence that contains all letters of the alphabet serves the same purpose. That means its decorative.

> This text has an insufficient contrast but it is decorative so it still passes.

Reviewer 2: I agree that  the meaning of the text is irrelevant here. But at the same time I feel a bit inconfortable saying that this can have very poor contrast and readability because
```html
<p>Helvetica is a widely used sans-serif typeface developed in 1957 by Max Miedinger and Eduard Hoffmann.</p>
<p style="font-family: helvetica; background: #000; color: #001;" aria-hidden="true">
	The quick brown fox jumps over the lazy dog.
</p>
```
is not really demonstrating the font (I can't see what it look like if I can't see the text…)

> This text has an insufficient contrast but it is decorative so it still passes.

Reviewer 6: I agree with @Jym77 here. Though the text can be completely replaced with another sentence, it is also a visible example of the font type. One could argue that is not strictly only an aesthetic purpose or no functionality, because it demonstrates the font to the user.

> This text has an insufficient contrast but it is decorative so it still passes.

Author: @Jym77 @DagfinnRomen  This aria-hidden block falls under 1.4.11: non-text contrast. It needs to meet 3:1, which in our example it does. Yes if it had a small contrast ratio, that would be a failure, but not of the SC we're testing in this rule.

> This text has an insufficient contrast but it is decorative so it still passes.

Reviewer 2: Hum, it's a bit hard to swallow that a text node is "non-text content", but I see your point and I agree with you.

It could be useful to expand a bit the description of the example to sum up this discussion (namely that the **meaning** is not important, hence it is not considered as text (it does not convey information) and thus is matched against 1.4.11, not 1.4.6).

```html
<p>Helvetica is a widely used sans-serif typeface developed in 1957 by Max Miedinger and Eduard Hoffmann.</p>
<p style="font-family: helvetica; background: #EEE; color: #777;" aria-hidden="true">
	The quick brown fox jumps over the lazy dog.
</p>
```

#### Passed Example 8

This text has an insufficient contrast but it does not express anything in human language so it still passes.

```html
<p style="color: #000; background: #666;">

> 	----=====++++++++___________***********%%%%%%%%%%%±±±±@@@@@@@@

Reviewer 2: I'm still challenging a bit the definition of "human language".
I think something like that would fail:
```html
Can you decipher the secret code here?
<p style="color: #000; background: #666;">
	----=====++++++++___________***********%%%%%%%%%%%±±±±@@@@@@@@
</p>
```
The point here being that this still doesn't really express anything in a human language, but it is important that it has good contrast. Additionally, the same element could pass or fail depending on context.

As far as I can tell, WCAG says nothing about human language but has "For example, if random words are used to create a background and the words could be rearranged or substituted without changing meaning". Would "words (or charcters?) can be rearranged without altering meaning" be a better formulation of the exception? (that would pass the garbage alone but fail it if part of a cipher (unless this is an elaborate cipher where characters order doesn't matter...))

Reviewer 3: I would state that a secret code is definitly a human language.

In a different context this specific element would indeed fail. But because that specific context is missing in this example we pass it.

Reviewer 2: If context is important and elements cannot be checked individually in isolation, we should probably add a note about it.

Reviewer 6: Strictly speaking, as "----=====++++++++___________***********%%%%%%%%%%%±±±±@@@@@@@@" is not written in a human language, this would not be considered text according to WCAG.

"Text: sequence of characters that can be programmatically determined, where the sequence is expressing something in human language".

I am aware that the definition of text differs from "text node", which is used in this rule. However, I would suggest moving this test case to inapplicable,

Author: Agree with Dagfinn. This isn't human language, even if it's a code. It may require a text alternative (unless it's a test, see 1.1.1). There is no scenario in which 1.4.3 applies, but 1.4.11 might in case of a code.

Reviewer 2: I still feel that the definition of Human Language (in WCAG) is open-ended enough to allow cipher ("_language that is spoken, written or signed (through visual or tactile means) to communicate with humans_") and that it is not clear when a secret code becomes a human language or not (for example, pig latin or Templar cipher are widely known and I would say they are human languages, otoh it's not clear that what's written on Phaistos Disc is Human language even though it might have been super clear for people of that time…)

But that's OK, I'm happy with the discussion and the explanation and won't push it further 😄

</p>
```

#### Passed Example 9

> This text is in a shadow DOM tree and has sufficient contrast.

Reviewer 5: I believe it would be more useful to say something like "Even though the text is in a shadow DOM tree, the text node is a child of an HTML in the flat tree and has sufficient contrast"

```html
<p style="color: #333; background: #fff;" id="p"></p>
<script>
	const shadowRoot = document.getElementById('p').attachShadow({ mode: 'open' })
	shadowRoot.textContent = 'Some text in English'
</script>
```

#### Passed Example 10

> This black [text node](https://dom.spec.whatwg.org/#text) is in Japanese, showing that this rule applies to any human language. (The phrase means "Even monkeys fall from trees.").

Reviewer 5: I agree with @Brynanders

> This black [text node](https://dom.spec.whatwg.org/#text) is in Japanese, showing that this rule applies to any human language. (The phrase means "Even monkeys fall from trees.").

Reviewer 3: You have convinced us. Despite the fact that I loved this expression. :(

```html
<p style="color: #333; background: #FFF;">
	猿も木から落ちる。
</p>
```

### Failed

#### Failed Example 1

> This text has an insufficient contrast with the plain background.

Reviewer 5: plain > white

```html
<p style="color: #AAA; background: white;">
	Some text in English
</p>
```

#### Failed Example 2

This text has an insufficient contrast with the gradient background.

```html
<p style="color: #AAA; background: linear-gradient(to right, #FFF, #00F);">
	Some text in English
</p>
```

#### Failed Example 3

This text has an insufficient contrast with the background image.

```html
<p
	style="color: #555; height:50px; padding-top:25px; background: black no-repeat -20px -20px url('../test-assets/contrast/black-hole.jpeg');"
>
	Black hole sun
</p>
```

#### Failed Example 4

This text fails because of alpha transparency.

```html
<p style="color: rgba(0,0,0,.3); background: #FFF">
	Some text in English
</p>
```

#### Failed Example 5

This text fails because of CSS transparent.

```html
<div style="background: #FFF">
	<p style="color: #000; opacity: .3">
		Some text in English
	</p>
</div>
```

#### Failed Example 6

> This text is in a shadow DOM tree and has insufficient contrast.

Reviewer 5: Update to mention that the text node is in the flattened tree

```html
<p style="color: #aaa; background: #fff;" id="p"></p>
<script>
	const shadowRoot = document.getElementById('p').attachShadow({ mode: 'open' })
	shadowRoot.textContent = 'Some text in English'
</script>
```

### Inapplicable

> #### Inapplicable Example 1

Reviewer 3: Yes. This is why this is an inapplicable example. ;)

>
	Black hole sun
</p>
```

#### Passed Example 4

This text passes because the text shadow gives the text sufficient contrast.

```html
<p style="color: #000; background: #737373; text-shadow: white 0 0 3px">
	Some text in a human language
</p>
```

#### Passed Example 5

This text only has a 3:1 contrast but it is a large (18 point) text so it still passes.

```html
<p style="color: #000; font-size:18pt; background: #666;">
	Some text in a human language
</p>
```

#### Passed Example 6

This text only has a 3:1 contrast but it is a large (14 points and bold) text so it still passes.

```html
<p style="color: #000; font-size:14pt; font-weight:700; background: #666;">
	Some text in English
</p>
```

#### Passed Example 7

This text has an insufficient contrast but it is decorative so it still passes.

```html
<p>Helvetica is a widely used sans-serif typeface developed in 1957 by Max Miedinger and Eduard Hoffmann.</p>
<p style="font-family: helvetica; background: #EEE; color: #777;" aria-hidden="true">
	The quick brown fox jumps over the lazy dog.
</p>
```

#### Passed Example 8

> This text has an insufficient contrast but it does not express anything in human language so it still passes.

Reviewer 2: This week-end I stumbled upon an interesting case for the "no human language" exception…
Consider this page: https://genius.com/Ishtar-band-o-julissi-lyrics
It's main content is the lyrics of the song _O julissi_ which "is sung entirely in Ukrainian-influenced imaginary language" [Wikipedia](https://en.wikipedia.org/wiki/O_Julissi)

Now, this does not express anything in human language, and was meant so. But I feel that the lyrics should be considered as text and have high contrast because they need to be readable…

Would you consider that:
- this is text content and requires 4.5:1 contrast?
- this is non-text content and requires 3:1 contrast?
- this does not express anything in human language and has no contrast requirement?

Author: I don't think this is relevant to the rule.

Reviewer 6: So this is where Eurovision and WCAG collide :) The same case could be made for these lyrics: https://genius.com/Urban-trad-sanomi-eurovision-edit-lyrics (A far superior song. Also Belgian, make of that what you will).

Either way, I agree with @WilcoFiers here. I would not count these lyrics as written in a human language and this is where WCAG draws the line.

accessibility_requirements:
  wcag20:1.4.3: # Contrast (Minimum)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed

input_aspects:
  - DOM Tree
  - CSS Styling
  - Language

authors:
  - Brian Bors
  - Kasper Isager
  - Wilco Fiers
---

## Applicability

Any [visible](#visible) character in a [text node](https://dom.spec.whatwg.org/#text) that is a [child](https://dom.spec.whatwg.org/#concept-tree-child) (in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree)) of an HTML element, except if the [text node](https://dom.spec.whatwg.org/#text) is a [descendant](https://dom.spec.whatwg.org/#concept-shadow-including-descendant) of an element that:

- Has a [semantic role](#semantic-role) that inherits from [widget](https://www.w3.org/TR/wai-aria-1.1/#widget)
- Is used in the [accessible name](#accessible-name) of a [widget](https://www.w3.org/TR/wai-aria-1.1/#widget) that is [disabled]()
- Has a [semantic role](#semantic-role) of [group](https://www.w3.org/TR/wai-aria-1.1/#group) that is [disabled]()

> **Note**: When the text color is the same as the background color, the element is not visible, and so is not applicable.

Reviewer 2: Technically in that case, the **element** is visible, but its text node isn't.
```html
<div style="background:white"><span style="background:red; color:red">Foo</span></div>
```
The `span` is visible (it is a red rectangle, and making it transparent changes the rendered pixels). The text inside the `span` is not visible (it is red on red, making it transparent will not change anything).
This is not a problem for this rule (because it does already target individual characters in the text nodes), but this note should nonetheless be made correct.

Making the element transparent:
```html
<div style="background:white"><span style="background:red; color:red; opacity:0">Foo</span></div>
```

Making the text transparent:
```html
<div style="background:white"><span style="background:red; color:rgba(1,0,0,0)">Foo</span></div>
```

---
id: afw4f7
name: Text node has minimum contrast
rule_type: atomic

description: |
  This rule checks that the highest possible contrast of every text node with its background meets the minimal contrast requirement

accessibility_requirements:
  wcag20:1.4.3: # Contrast (Minimum)
    forConformance: true
    failed: not satisfied
    passed: further testing needed

>     inapplicable: further testing needed

Reviewer 2: This rule should also map to 1.4.6 (Contrast (enhanced)).
Essentially, failing 1.4.1 also fails 1.4.6. So if this rule fails, then 1.4.6 fails.

Reviewer 6: I see your point, @Jym77 , though I think it is tidier to map this only to 1.4.3.

What do we usually do in these cases?

Reviewer 2: I am not sure the case already arose in published rules. I can see at least two other PR that would be affected by a decision like that:
- #555: if the link (in context) is not descriptive, then the link (alone) is not descriptive. @carlosapaduarte chose to not have 2.4.9 as extra requirement.
- #1027: if the element has no parent with a language attribute, then the page itself has none. @patrickhlauke and me had some discussion about adding 3.1.1 as extra requirement: https://github.com/act-rules/act-rules.github.io/pull/1027#discussion_r349030552 and following

Both for this PR and for #555, it does make sense to write another rule for the AAA criterion afterwards. Even if AAA is not our main focus, these will essentially be "free" rules given that all the difficult stuff has already been discussed. For #1027, the "global" language rules already exist.

---
It seems that in each of these 3 cases, the rule writers decided to only map to the more restrictive criterion… So maybe it is indeed the way we should do 😄 
My idea was that the extra mapping is for free and does give more information to users and more "punch" to the rules…

Reviewer 2: As per [ACT rule format](https://www.w3.org/TR/act-rules-format/#accessibility-requirements-mapping):
> the rule must list all accessibility requirements from those documents that are not satisfied when one or more of the outcomes of the rule is failed.

So I'd say that the format is actually requiring that we put both (here, adding 1.4.6).

Reviewer 2: Hmmm… At the same time we have the habit that "passed examples must pass the SC", so if the rule also maps to 1.4.6 then all passed examples must have enhanced contrast… 🤔 

However, that habit is not part of the ACT rule format (as far as I can tell), while listing all requirements that fail when the rule fails is…

input_aspects:
  - DOM Tree
  - CSS Styling
  - Language

acknowledgements:
  authors:
    - Brian Bors
    - Kasper Isager
    - Wilco Fiers
---

## Applicability

> Any [visible](#visible) character in a [text node](https://dom.spec.whatwg.org/#text) that is a [child](https://dom.spec.whatwg.org/#concept-tree-child) (in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree)) of an HTML element, except if the [text node](https://dom.spec.whatwg.org/#text) is a [descendant](https://dom.spec.whatwg.org/#concept-shadow-including-descendant) of an element that:

Reviewer 5: "Any visible character in a text node" means that the expectation will apply to the character, not to the text node. I'm pretty sure that this is not what you intended, otherwise a lot of passed examples fail because some of the characters do not meet the expectation.

Suggestion: Any text node with visible characters, that is a child ...

- Has a [semantic role](#semantic-role) that inherits from [widget](https://www.w3.org/TR/wai-aria-1.1/#widget)
- Is used in the [accessible name](#accessible-name) of a [widget](https://www.w3.org/TR/wai-aria-1.1/#widget) that is [disabled]()
- Has a [semantic role](#semantic-role) of [group](https://www.w3.org/TR/wai-aria-1.1/#group) that is [disabled]()

**Note**: When the text color is the same as the background color, the [text node](https://dom.spec.whatwg.org/#text) is not [visible](#visible), and so it does not need to be tested for contrast.

## Expectation

For each test target, the [highest possible contrast](#highest-possible-contrast) between the [foreground colors](#foreground-colors-of-text) and [background colors](#background-colors-of-text) is at least 4.5:1 or 3.0:1 for [larger scale text](#large-scale-text), except if the [text node](https://dom.spec.whatwg.org/#text) is [decorative](#decorative), or doesn't express anything in [human language](https://www.w3.org/TR/WCAG21/#dfn-human-language-s).

> **Note**: Passing this rule does not mean that the text has sufficient color contrast. If all background pixels have a low contrast with all foreground pixels, the success criterion is guaranteed to not be satisfied. When some pixels have sufficient contrast, and others do not, ledgability should be considered. There is no clear method for determining ledgability, which is why this is out of scope for this rule.

Reviewer 6: ```suggestion
**Note**: Passing this rule does not mean that the text has sufficient color contrast. If all background pixels have a low contrast with all foreground pixels, the success criterion is guaranteed to not be satisfied. When some pixels have sufficient contrast, and others do not, legibility should be considered. There is no clear method for determining legibility, which is why this is out of scope for this rule.
```

#### Passed Example 2

This dark grey text is on a partially white gradient background.

```html
<p style="color: #333; background: linear-gradient(to right, #FFF, #00F); width: 300px;">
	Some text in a human language
</p>
```

#### Passed Example 3

This light grey text is on a partially black background image.

```html
<p
	style="color: #CCC; height:50px; padding-top:25px; background: #000 no-repeat -20px -20px url('../test-assets/contrast/black-hole.jpeg');"
>
	Black hole sun
</p>
```

#### Passed Example 4

This text passes because the text shadow gives the text sufficient contrast.

```html
<p style="color: #000; background: #737373; text-shadow: white 0 0 3px">
	Some text in a human language
</p>
```

#### Passed Example 5

This text only has a 3:1 contrast but it is a large (18 point) text so it still passes.

```html
<p style="color: #000; font-size:18pt; background: #666;">
	Some text in a human language
</p>
```

#### Passed Example 6

This text only has a 3:1 contrast but it is a large (14 points and bold) text so it still passes.

```html
<p style="color: #000; font-size:14pt; font-weight:700; background: #666;">
	Some text in English
</p>
```

#### Passed Example 7

> The purpose of the helvetica letters is to show what the font looks like. The actual words are irrelevant and could be replaced with any other sentence that includes all letters of the alphabat. Because of that, the rule is passed, despite the text not meeting the 4.5:1 color contrast requirement.

Reviewer 5: ```suggestion
The purpose of the Helvetica letters is to show what the font looks like. The actual words are irrelevant and could be replaced with any other sentence that includes all letters of the alphabet. Because of that, the rule passes, despite the text not meeting the 4.5:1 color contrast requirement.
```

> **Note**: Because this is non-text content, [success criterion 1.4.11 Non-text Contrast] requires font example to have a color contrast of 3:1.

Reviewer 5: The link to the SC is missing

```html
<p>Helvetica is a widely used sans-serif typeface developed in 1957 by Max Miedinger and Eduard Hoffmann.</p>
<p style="font-family: helvetica; background: #EEE; color: #777;" aria-hidden="true">
	The quick brown fox jumps over the lazy dog.
</p>
```

#### Passed Example 8

This text has an insufficient contrast but it does not express anything in human language so it still passes.

```html
<p style="color: #000; background: #666;">
	----=====++++++++___________***********%%%%%%%%%%%±±±±@@@@@@@@
</p>
```

#### Passed Example 9

> Even though the text is in a shadow DOM tree, the text is part of a child [text node](https://dom.spec.whatwg.org/#text) of an HTML in the flat tree and has sufficient contrast

Reviewer 5: ```suggestion
Even though the text is in a shadow DOM tree, the text is part of a child [text node](https://dom.spec.whatwg.org/#text) of an HTML element in the flat tree and has sufficient contrast
```

```html
<p style="color: #333; background: #fff;" id="p"></p>
<script>
	const shadowRoot = document.getElementById('p').attachShadow({ mode: 'open' })
	shadowRoot.textContent = 'Some text in English'
</script>
```

### Failed

#### Failed Example 1

> This text has an insufficient contrast with the white background.

Reviewer 5: ```suggestion
This text has insufficient contrast with the white background.
```

```html
<p style="color: #AAA; background: white;">
	Some text in English
</p>
```

#### Failed Example 2

> This text has an insufficient contrast with the darkest point on the gradient background behind the text.

Reviewer 5: ```suggestion
This text has insufficient contrast with the darkest point on the gradient background behind the text.
```

```html
<p style="color: #AAA; background: linear-gradient(to right, #FFF, #00F); width: 300px">
	Some text in English
</p>
```

#### Failed Example 3

> This text has an insufficient contrast with the background image.

Reviewer 5: ```suggestion
This text has insufficient contrast with the background image.
```

```html
<p
	style="color: #555; height:50px; padding-top:25px; background: black no-repeat -20px -20px url('../test-assets/contrast/black-hole.jpeg');"
>
	Black hole sun
</p>
```

#### Failed Example 4

This text fails because the alpha transparency significantly lowers the contrast of the otherwise black text.

```html
<p style="color: rgba(0,0,0,.3); background: #FFF">
	Some text in English
</p>
```

#### Failed Example 5

This text fails because the CSS opacity property significantly lowers the contrast of the otherwise black text.

```html
<div style="background: #FFF">
	<p style="color: #000; opacity: .3">
		Some text in English
	</p>
</div>
```

#### Failed Example 6

> Even though the text is in a shadow DOM tree, the text is part of a child [text node](https://dom.spec.whatwg.org/#text) of an HTML in the flat tree and has insufficient contrast

Reviewer 5: ```suggestion
Even though the text is in a shadow DOM tree, the text is part of a child [text node](https://dom.spec.whatwg.org/#text) of an HTML element in the flat tree and has insufficient contrast
```

```html
<p style="color: #aaa; background: #fff;" id="p"></p>
<script>
	const shadowRoot = document.getElementById('p').attachShadow({ mode: 'open' })
	shadowRoot.textContent = 'Some text in English'
</script>
```

#### Failed Example 7

> The text is placed over a black and white background, The grey text fails because of its opacity, where the text on the black background becomes too dark, and the text on the white background becomes too light.

Reviewer 5: ```suggestion
The text is placed over a black and white background. The grey text fails because of its opacity, where the text on the black background becomes too dark, and the text on the white background becomes too light.
```

#### Passed Example 9

This text is in a shadow DOM tree and has sufficient contrast.

```html
<p style="color: #333; background: #fff;" id="p"></p>
<script>
	const shadowRoot = document.getElementById('p').attachShadow({ mode: 'open' })
	shadowRoot.textContent = 'Some text in English'
</script>
```

#### Passed Example 10

> This black [text node](https://dom.spec.whatwg.org/#text) is in Japanese, showing that this rule applies to any human language. (The prase means "Even monkeys fall from trees.").

Reviewer 2: I'm still understanding the sentence (and example) as "the rule applies to the primary language of the page, whatever it is" rather than "the rule applies to every text on the page, no matter what language it's in, even if it is different from the primary language of the page".

Imho, the example to show that would be (keeping the same idea) a page in English listing Japanese sayings (and their translations) (only one being actually needed in the example) and thus including some text in Japanese. If the English text has good contrast but the Japanese doesn't, then the failure of the rule shows that it is concerned about any text on the page, not only text in the primary language.
(if both Japanese and English text have good contrast, then it is still possible to believe that the rule does not care about the Japanese part)
